nifi-api/src/main/java/org/apache/nifi/migration/PropertyConfiguration.java

```diff
@@ -128,10 +128,83 @@ default Optional<String> getPropertyValue(PropertyDescriptor descriptor) {
+     * already exists within the same scope and with the same implementation and configuration, a new service may not be created and instead
+     * the existing service may be used.
+     * </p>
+     *
+     * <p>
+     * This allows for properties that were previously defined in the extension to be moved to a Controller Service. For example,
+     * consider a Processor that has "Username" and "Password" properties. In the next version of the Processor, we want to support
+     * multiple types of authentication, and we delegate the authentication to a Controller Service. Consider that the Controller Service
+     * implementation we wish to use has a classname of {@code org.apache.nifi.services.authentication.UsernamePassword}. We might then
+     * use this method as such:
+     * </p>
+     *
+     * <pre><code>
+     *     // Create a new Controller Service ofo type org.apache.nifi.services.authentication.UsernamePassword whose Username and Password
+     *     // properties match those currently configured for this Processor.
+     *     final Map&lt;String, String&gt; serviceProperties = Map.of("Username", propertyConfiguration.getRawPropertyValue("Username"),
+     *          "Password", propertyConfiguration.getRawPropertyValue("Password"));
+     *     final String serviceId = propertyConfiguration.createControllerService("org.apache.nifi.services.authentication.UsernamePassword", serviceProperties);
+     *
+     *     // Set our Authentication Service property to point to this new service.
+     *     propertyConfiguration.setProperty(AUTHENTICATION_SERVICE, serviceId);
+     *
+     *     // Remove the Username and Password properties from this Processor, since we are now going to use then Authentication Service.
+     *     propertyConfiguration.removeProperty("Username");
+     *     propertyConfiguration.removeProperty("Password");
+     * </code></pre>
+     *
+     * <p>
+     * Note the use of {@link #getRawPropertyValue(String)} here instead of {@link #getPropertyValue(String)}. Because we want to set
+     * the new Controller Service's value to the same value as is currently configured for the Processor's "Username" and "Password" properties,
+     * we use {@link #getRawPropertyValue(String)}. This ensures that if the Processor is configured using Parameters, those Parameter
+     * references are still held by the Controller Service.
+     * </p>
+     *
+     * <p>
+     * Also note that this method expects the classname of the implementation, not the classname of the interface.
+     * </p>
+     *
+     * @param implementationClassName the fully qualified classname of the Controller Service implementation
+     * @param serviceProperties       the property values to configure the newly created Controller Service with
+     * @return an identifier for the Controller Service
+     */
+    String createControllerService(String implementationClassName, Map<String, String> serviceProperties);
```

Comment:
The implementation can throw an `IllegalArgumentException`, but that makes sense in the end as this is an operation that must succeed based on requested implementation.

This ends up resulting a stronger requirement between a particular component and a Controller Service implementation, but as shown in the AWS use cases, that seems reasonable.

Reply:
Yeah, agreed. I don't love that it is needing to provide an explicit implementation of the service, but at the same time - the whole point is to allow backward compatibility and so it makes sense to declare that a specific implementation is what is necessary. Since it does not affect what is allowed to be configured, I think it's fair game.